### ISSUE TITLE
Write automation for the creation of S3 bucket to serve the React files

### ISSUE THREAD

**Issue by toddysm** [North-Seattle-College/ad440-winter2022-tuesday-repo#7]
### User Story
As a developer, I want to write a script that creates S3 buckets, so I don't have to do it manually every time

### Estimated time to complete
4 hours

### Acceptance Criteria
- [x] Create script that automatically generates a new S3 bucket
    - [x] Script should accept a name parameter for the bucket

### Pull Request
#18

**Comment by dansjack** [comment]
# PR Documentation
## Gallery

If your bucket name isn't unique, this is the error message you'll get when you run `aws cloudformation describe-stack-events --stack-name <<STACK_NAME>>`
<img width="600" alt="image" src="https://user-images.githubusercontent.com/30481844/149850255-fcab29fe-fa2e-46ba-9a72-1934e23d8a2b.png">

This shows the following:
1. navigating to the `devops` folder
2. attempting to run the command with a non-unique bucket name
3. reading the error logs
4. deleting the stack from the previous failed command `aws cloudformation delete-stack --stack-name <<STACK_NAME>>`
    * Note: `<<STACK_NAME>>` here begins with `arn` and ends with a string like `a36c6130-77f3-11ec-b26b-06300054abe1`
5. running the command successfully with a unique bucket name
<img width="600" alt="image" src="https://user-images.githubusercontent.com/30481844/149850533-53dc8a84-3c61-4db8-ae88-10e07dce3b29.png">

Example page served from S3 bucket
<img width="600" alt="image" src="https://user-images.githubusercontent.com/30481844/149851326-ddfbed2e-5e94-47dd-8d15-6d6da4b2e1c2.png">


## Testing

### Pre-reqs
* You must have the aws cli installed and it must be configured to use your access and secret key

### Steps
1. From the root of the repo, `cd devops`
2. Run the following script, replacing <<STACK_NAME>> and <<BUCKET_NAME>> with a unique name (I usually name them the same and add `-stack` and `-bucket` to the end)`aws cloudformation deploy --template-file ./makeReactBucket.json --stack-name <<STACK_NAME>> --parameter-overrides BucketName=<<BUCKET_NAME>>`
3. Once the stack is created, make sure the new stack and bucket have been created in the [S3](https://s3.console.aws.amazon.com/s3/home?region=us-west-2) and [Cloudformation](https://us-west-2.console.aws.amazon.com/cloudformation/home?region=us-west-2#/stacks?filteringStatus=active&filteringText=&viewNested=true&hideStacks=false) dashboards.
4. Copy/paste the following into a file called `index.html`:
```html
<!doctype html>

<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Public site</title>
</head>
<body>
  <h1>Hello, world!</hi>
</body>
</html>

```
5. In the Objects tab of your new S3 bucket, click upload and then choose the file you just created, then click upload again
6. Go to the Properties tab of the S3 bucket and scroll to the bottom. Go to the url for your Bucket website endpoint. If everything worked, you should see the Hello, world! text like in the Gallery above.

### Troubleshooting
I came across two issues during testing:
* Error when I gave a non-unique bucket name (solved by replacing with a unique name)
* Error when trying to update a stack that had failed to complete on the last attempt (solved by deleting the stack with `aws cloudformation delete-stack --stack-name`. See gallery screenshot 2 for more info

**Comment by dansjack** [comment]
# Activities Documentation


| Date  | Activity | Time Spent |
| ------------- | ------------- | ------------- |
| 01/17/2022 | Researching Cloudformation and S3s for serving static sites  | 2 hr  |
| 01/17/2022 | Prototyping Cloudformation template + npm script  | 1 hr  |
| 01/18/2022  | Testing | 0.5 hr  |
| 01/18/2022  | Documentation | 1 hr  |

**Comment by dansjack** [comment]
Closed via #18
